### ISSUE TITLE
PHPUnit "Bugs and best practices" inspection ignores aliases

### ISSUE THREAD

**Issue by schlessera** [kalessil/phpinspectionsea#110]
The PHPUnit inspection "Bugs and best practices" seems to ignore aliases and throws warnings when these are used.

Example code:
```php
use SomeClass as Testee;

class PHPUnit_Framework_TestCase {

	/**
	 * Tests something.
	 *
	 * @covers Testee::someMethod()
	 */
	public function test_something() {
		$result = Testee::seomMethod( $$argument );
	}
}
```

In the above example, PHPStorm will correctly recognize the `@covers Testee::someMethod()` (and even helps with auto-completion).

However, the above mentioned inspection then throws a warning on `test_something()`, stating that:
```
@covers referencing to a non-existing entity
```

It should recognize the alias and know that the class and its method do exist.

**Comment by kalessil** [comment]
Thank you for reporting @schlessera , I'll check if this an IDE or Php Inspections (EA Extended) issue.

**Comment by kalessil** [comment]
@schlessera : can not reproduce this: all tests, incl. the one similar to reported are passing.

**Comment by kalessil** [comment]
What happens if you pressing Ctrl and pointing to Testee and someMethod, are both become clickable?

**Comment by schlessera** [comment]
> What happens if you pressing Ctrl and pointing the cursor to Testee and someMethod, are both become clickable?

Yes, that works as expected.

The code where I was experiencing the problem is now directly accessible here on GitHub: https://github.com/pluginkollektiv/antispam-bee/blob/master/tests/Unit/AntispamBeeTest.php#L80

So, you can check out this code to see whether you get the same error there.

<img width="577" alt="screenshot 2017-02-04 18 52 08" src="https://cloud.githubusercontent.com/assets/83631/22620381/12b798e6-eb0b-11e6-9397-9ecc351d2c66.png">

In case it can help, PHPStorm and JDK versions:
<img width="324" alt="screenshot 2017-02-04 18 53 07" src="https://cloud.githubusercontent.com/assets/83631/22620388/368ff8a8-eb0b-11e6-8045-f1388a4ac3b5.png">

**Comment by kalessil** [comment]
The same PhpStorm, the same project and not reproducible.
I modified the logic a little bit, can you try installing PhpInspectionsEA.jar from the repository?

To install:
Download the file, then in IDE: File -> Settings -> Plugins and under the plugins list click *Install plugin from disk* and installing the downloaded file.

**Comment by kalessil** [comment]
ping @schlessera

**Comment by kalessil** [comment]
@schlessera : any feedback on the fix?

**Comment by schlessera** [comment]
Ah, sorry, let me try that right now...

**Comment by schlessera** [comment]
No, seems to still give the same result:

<img width="822" alt="screenshot 2017-02-08 17 26 23" src="https://cloud.githubusercontent.com/assets/83631/22746374/ca594fd8-ee23-11e6-9d9d-9bfb8bcc703c.png">

Is there a way for me to verify that I am indeed running the downloaded version of the plugin? I see version 2.3.1.1 here.

**Comment by kalessil** [comment]
IDE suggest restarting to get new versions loaded, that's a pretty reliable indicator.

If you don't mind I'll build a special debug version late today and ping you back for next debug round...

**Comment by schlessera** [comment]
Yes, I needed to restart the IDE.

Sure, I'm happy to help.

**Comment by kalessil** [comment]
Next round), same file need but from [this branch](https://github.com/kalessil/phpinspectionsea/tree/bugfix-110).

Please inspect the file and share what the inspection reports: messages starting with "debug: " are kind of echo 1 and etc, but gives an idea what happens on your end )

**Comment by schlessera** [comment]
<img width="311" alt="screenshot 2017-02-09 08 21 32" src="https://cloud.githubusercontent.com/assets/83631/22773130/cd407c12-eea0-11e6-8471-cd7a3aca0d4f.png">

**Comment by kalessil** [comment]
Emm, something goes completely wrong: you can navigate the method, but the code representation doesnot contain any recognized references.

How the doc block looks like now (I's expect the tag to be striked out)?

**Comment by schlessera** [comment]
Yes, the `@covers` line is striked out:

<img width="577" alt="screenshot 2017-02-09 09 08 29" src="https://cloud.githubusercontent.com/assets/83631/22774425/611f0d1c-eea7-11e6-823a-598c084977bb.png">

**Comment by kalessil** [comment]
Ok, then I need to comeup with extended solution for this case - will ping you back late evening today.

**Comment by schlessera** [comment]
Yes, let me know if you need anything.

Again, if you want to have a testing ground, you can checkout the code that causes this: https://github.com/pluginkollektiv/antispam-bee/blob/master/tests/Unit/AntispamBeeTest.php#L80

**Comment by kalessil** [comment]
I have it ;), the problem is that my environment is not triggering the issue and I'm lucky that it's reproduciable in yourth.

**Comment by schlessera** [comment]
Maybe a conflict with another plugin?

**Comment by kalessil** [comment]
Probably, it's almost impossible to debug such things.

**Comment by schlessera** [comment]
I just now reduced my enabled plugins to a minimum, and the issue is still there.

These are the plugins that are still active now:
<img width="657" alt="screenshot 2017-02-09 10 13 31" src="https://cloud.githubusercontent.com/assets/83631/22776585/7c385320-eeb0-11e6-888f-a8d4ca6c8316.png">

**Comment by kalessil** [comment]
Looks good: no framework integrations and etc. Strange...

**Comment by kalessil** [comment]
@schlessera : fixed (a fresh binary in the master already).
I have no clue what the root cause but provided a solution + corresponding UTs.

**Comment by schlessera** [comment]
@kalessil Unfortunately, the bug is still not fixed on my end:

<img width="1076" alt="screenshot 2017-02-10 08 41 25" src="https://cloud.githubusercontent.com/assets/83631/22818306/dca1ce1e-ef6c-11e6-9660-f0afa5acabb2.png">

**Comment by kalessil** [comment]
That's weird, the plugin attempts now to find the method by it's own.
Will be a skype call tomorrow an option for you?

**Comment by schlessera** [comment]
I'm running behind on projects, atm. Skype call would be okay starting mid next week.

**Comment by kalessil** [comment]
Ok, then I'll ping you mid of the next week.

**Comment by kalessil** [comment]
@schlessera how is it going, any chance to skype today evening?

**Comment by schlessera** [comment]
Sorry, @kalessil, too much going on atm. I'll try to recontact you about this as soon as things have calmed down.

**Comment by kalessil** [comment]
No problems @schlessera

**Comment by b1rdex** [comment]
@schlessera have you tested phpunit with coverage enabled for such scenario? Phpunit says it can't find class when I try to cover aliased name. As I know, `@covers` must have FQN class name + optionally method.
```
anatoly@zapil /tmp/phpunit {master} $ cat tests/AliasTest.php
<?php

use Acme\Tax;
use Acme\Tax as Alias;

use PHPUnit\Framework\TestCase;

class AliasTest extends TestCase
{
        /**
         * @covers Acme\Tax::__construct()
         */
        public function testByName()
        {
                $sut = new Tax();
        }

        /**
         * @covers Alias::__construct()
         */
        public function testByAlias()
        {
                $sut = new Alias();
        }
}
anatoly@zapil /tmp/phpunit {master} $ vendor/bin/phpunit --coverage-text=coverage --whitelist=src/ tests/AliasTest.php
PHPUnit 5.7.15 by Sebastian Bergmann and contributors.

.W                                                                  2 / 2 (100%)

Time: 43 ms, Memory: 4.00MB

There was 1 warning:

1) AliasTest::testByAlias
Trying to @cover or @use not existing method "Alias::__construct".

WARNINGS!
Tests: 2, Assertions: 0, Warnings: 1.
```

I think issue should be closed and that inspection works as it should.

**Comment by b1rdex** [comment]
I created test repo: https://github.com/b1rdex/phpunit-covers-alias
Note you need php-xdebug extension installed to run coverage.
```sh
cd /tmp
git clone https://github.com/b1rdex/phpunit-covers-alias
cd phpunit-covers-alias
composer install
vendor/bin/phpunit --coverage-text=coverage --whitelist=src/ tests/AliasTest.php
```

**Comment by kalessil** [comment]
Ok, closing then.
